File: devices-api-java-mvc/src/main/java/com/lucas/devicesapijavamvc/kafka/DeviceEventListener.java
package com.lucas.devicesapijavamvc.kafka;

import com.lucas.devicesapijavamvc.domain.Device;
import com.lucas.devicesapijavamvc.domain.DeviceState;
import com.lucas.devicesapijavamvc.repository.DeviceRepository;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.stereotype.Component;

@Component
@ConditionalOnProperty(value = "kafka.enabled", havingValue = "true")
public class DeviceEventListener {

    private final DeviceRepository repo;

    public DeviceEventListener(DeviceRepository repo) {
        this.repo = repo;
    }

    @KafkaListener(
            topics = "${kafka.topic.devices}",
            groupId = "java-devices",
            containerFactory = "deviceEventKafkaListenerFactory"
    )
    public void onDeviceEvent(DeviceEvent event) {
        try {
            if ("JAVA".equals(event.origin())) {
                return; // ignora eventos próprios
            }

            var existingOpt = repo.findByExternalId(event.externalId());
            if (existingOpt.isPresent()) {
                var existing = existingOpt.get();
                existing.setName(event.name());
                existing.setBrand(event.brand());
                existing.setState(DeviceState.valueOf(event.state()));
                repo.save(existing);
            } else {
                Device mirrored = Device.fromEvent(
                        event.externalId(),
                        event.name(),
                        event.brand(),
                        DeviceState.valueOf(event.state()),
                        event.creationTime()
                );
                repo.save(mirrored);
            }
        } catch (Exception ex) {
            // log simples para challenge
            System.err.println("Failed to process DeviceEvent in Java: " + event);
            ex.printStackTrace();
        }
    }
}

File: devices-api-kotlin-reactive/src/main/kotlin/com/lucas/devicesapikotlinreactive/infrastructure/messaging/DeviceEventListener.kt
package com.lucas.devicesapikotlinreactive.infrastructure.messaging

import com.lucas.devicesapikotlinreactive.application.service.DeviceService
import com.lucas.devicesapikotlinreactive.domain.model.Device
import com.lucas.devicesapikotlinreactive.domain.model.DeviceState
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono

@Component
@ConditionalOnProperty(value = ["kafka.enabled"], havingValue = "true")
class DeviceEventListener(
    private val service: DeviceService
) {

    private val log = LoggerFactory.getLogger(javaClass)

    @KafkaListener(
        topics = ["\${kafka.topic.devices}"],
        groupId = "kotlin-devices",
        containerFactory = "deviceEventKafkaListenerFactory"
    )
    fun onDeviceEvent(event: DeviceEvent) {
        try {
            if (event.origin == "KOTLIN") return

            val device = Device.fromEvent(
                externalId = event.externalId,
                name = event.name,
                brand = event.brand,
                state = DeviceState.valueOf(event.state),
                creationTime = event.creationTime
            )


            service.upsertByExternalId(device)
                .onErrorResume {
                    log.error("Failed to upsert device from event: $event", it)
                    Mono.empty()
                }
                .subscribe()

        } catch (ex: Exception) {
            log.error("Failed to process DeviceEvent in Kotlin: $event", ex)
        }
    }
}
